### PR TITLE
add colocation identifier

### PR DESCRIPTION
**Pull request by nicolai86**
## What does this PR do?

Expose `colocation_group_id`; this will allow users of the lightning platform to use high speed interconnects.
As this has a couple of moving pieces I'll circle back and add documentation once this feature works end to end in production.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/lightning/app/utilities/packaging/cloud_compute.py [hunk 1]
```diff
@@ -81,6 +81,9 @@ class CloudCompute:
 
         mounts: External data sources which should be mounted into a work as a filesystem at runtime.
 
+        colocation_group_id: Identifier for groups of works to be colocated in the same datacenter.
```
Comment (tchaton) [inline src/lightning/app/utilities/packaging/cloud_compute.py:84]: Love it !

**Review by tchaton — APPROVED** [review]
LGTM !
 Maybe a test ;)

File: src/lightning/app/utilities/packaging/cloud_compute.py [hunk 2]
```diff
@@ -81,6 +81,9 @@ class CloudCompute:
...
+            Can be a string of max. 64 characters.
```
Comment (awaelchli) [inline src/lightning/app/utilities/packaging/cloud_compute.py:85]:
```suggestion
            Set this to a string of max. 64 characters and all works with this group id will run in the same datacenter.
            If not set, the works are not guaranteed to be colocated.
```
@nicolai86 Just a suggestion to make it a bit more explicit that the user chooses a name. I'm not sure about the last sentence, it's just what I guessed would be the default behavior.

File: src/lightning/app/utilities/packaging/cloud_compute.py [hunk 3]
```diff
@@ -118,6 +122,9 @@ def __post_init__(self) -> None:
         if self._internal_id is None:
             self._internal_id = self._generate_id()
 
+        if isinstance(self.colocation_group_id, str) and len(self.colocation_group_id) > 64:
```
Comment (awaelchli) [inline src/lightning/app/utilities/packaging/cloud_compute.py:125]:
```suggestion
        if not isinstance(self.colocation_group_id, str) or len(self.colocation_group_id) > 64:
```
Isn't this what we want to check?

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@nicolai86 Some tests need adjustment because of the new attribute in the state.

**Comment by Borda** [comment]
@awaelchli could you help here to finish it?

**Comment by awaelchli** [comment]
@nicolai86 Should this work be continued?

**Comment by lantiga** [comment]
Yes we are going to need it as a general attribute.

**Review by Borda — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

**Pull request merged**